sstables/sstable_set.cc

```diff
@@ -1406,6 +1406,29 @@ mutation_reader sstable_set::make_crawling_reader(
     return make_combined_reader(schema, std::move(permit), std::move(readers), streamed_mutation::forwarding::no, mutation_reader::forwarding::no);
 }
 
+mutation_reader sstable_set::make_streaming_reader(
+        schema_ptr schema,
+        reader_permit permit,
+        const dht::partition_range& range) {
+    auto trace_state = tracing::trace_state_ptr();
+    if (range.is_full()) {
+        return make_crawling_reader(
+            std::move(schema),
+            std::move(permit),
+            std::move(trace_state),
+            sstables::default_read_monitor_generator());
+    } else {
+        auto& slice = schema->full_slice();
+        return make_range_sstable_reader(
+            std::move(schema),
+            std::move(permit),
+            range,
+            std::move(trace_state),
+            streamed_mutation::forwarding::no,
+            mutation_reader::forwarding::no);
+    }
```

Comment:
I think this could be simply folded into the regular method which creates the reader.
We can use the crawling reader every time the caller wants to read the full range, e.g. for compaction.

Reply:
Yes, we also need to check the slice. Indeed there is no good way to check this currently, although it is not hard to do:
```
!slice.get_specific_ranges() && slice.default_row_ranges().size() == 1 && slice.default_row_ranges().front().is_full();
```

We should add an `is_full()` method to the partition slice to avoid repeating this logic elsewhere.
Maybe the method should check that all columns are requested too, although the sstable reader unconditionally includes all columns anyway, in fact all readers do. Unwanted columns are dropped when the query result is built.

Reply:
to check if all columns are requested, we need to check if the `static_columns` and `regular_columns` contains the same set as the ones in the table's schema. and when `make_reader()` is called, we do have access to the schema, so it's possible to check the columns as well. but i think it'd be much more complicated. as we would need to compare two sets stored in a vector. we could assume the order, but this contract is not enforced anywhere. so it could be fragile. so in the latest revision, i am not checking these member variables.

Reply:
Like I said above, all readers ignore this, so it is fine. You only need to check the ranges and specific ranges.

Reply:
@denesb turns out existing callers are skipping the partitions even they asked for a full-ranged reader. see https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10993/#showFailuresLink . so i am reverting the change which updates `make_reader()` to use the full-scan reader if the range and slice are full, and returning to the previous approach of https://github.com/scylladb/scylladb/pull/20187#issuecomment-2295840029